### PR TITLE
Implement federated metric metadata API

### LINKED ISSUE #3345: Implement metrics metadata API
Currently multiple UI tools (such as PromLens and Grafana) surface metric metadata (HELP + TYPE information) to the user when building queries. This is very useful, but can only work when the PromQL-compatible endpoint also implements the metric metadata API endpoint (https://prometheus.io/docs/prometheus/latest/querying/api/#querying-metric-metadata). Thanos itself does not store this information for historical metrics yet, but it would be possible to obtain current metadata from Prometheus servers via a new feature in the Thanos Sidecar + Store protocol and then aggregate them into a response to the Query client. Would this be doable or am I missing something?

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: Ben Ye <yb532204897@gmail.com>

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

Fixes #3345 

This pr follows the awesome pr https://github.com/thanos-io/thanos/pull/3350 and comment https://github.com/thanos-io/thanos/issues/3345#issuecomment-737172268 to implement the Prometheus metric metadata API https://prometheus.io/docs/prometheus/latest/querying/api/#querying-metric-metadata.

## Verification

Tested manually and E2E tests added.

### CODE REVIEW

**Comment by bwplotka** [comment]
I think this slipped through notification - this is greater, let's Review! 

cc @lilic (in regards to Info API re-design), @kakkoyun @brancz @squat

File: pkg/query/storeset.go [hunk 1]
```diff
@@ -554,6 +586,20 @@ func (s *StoreSet) GetRulesClients() []rulespb.RulesClient {
 	return rules
 }
 
+// GetRulesClients returns a list of all active rules clients.
```
Comment (yeya24) [inline pkg/query/storeset.go:589]: This comment needs to be updated.

File: pkg/query/storeset.go [hunk 2]
```diff
@@ -49,6 +50,11 @@ type RuleSpec interface {
 	Addr() string
 }
 
+type MetadataSpec interface {
+	// Addr returns RulesAPI Address for the rules spec. It is used as its ID.
```
Comment (yeya24) [inline pkg/query/storeset.go:54]: This needs to be updated.

File: cmd/thanos/query.go [hunk 3]
```diff
@@ -399,6 +432,9 @@ func runQuery(
 				if err := dnsRuleProvider.Resolve(resolveCtx, ruleAddrs); err != nil {
 					level.Error(logger).Log("msg", "failed to resolve addresses for rulesAPIs", "err", err)
 				}
+				if err := dnsMetadataProvider.Resolve(resolveCtx, metadataAddrs); err != nil {
+					level.Error(logger).Log("msg", "failed to resolve addresses for rulesAPIs", "err", err)
```
Comment (yeya24) [inline cmd/thanos/query.go:436]: The content needs to be updated.

File: cmd/thanos/query.go [hunk 4]
```diff
@@ -141,6 +148,10 @@ func registerQuery(app *extkingpin.App) {
 			return errors.Errorf("Address %s is duplicated for --rule flag.", dup)
 		}
 
+		if dup := firstDuplicate(*metadataEndpoints); dup != "" {
+			return errors.Errorf("Address %s is duplicated for --target flag.", dup)
```
Comment (yeya24) [inline cmd/thanos/query.go:152]: The content needs to be updated.

**Comment by kakkoyun** [comment]
@yeya24 is this ready for another round of review?

Also it worths having a look: 
CleanShot 2021-02-12 at 11 26 27@2x

**Comment by yeya24** [comment]
@kakkoyun Thanks! I have fixed that issue and this pr is ready for review.

**Comment by juliusv** [comment]
Reminder poke to the reviewers (this feature would be awesome to have) :)

File: test/e2e/metadata_api_test.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+		defaultPromConfig("ha", 0, "", "", "localhost:9090", "sidecar-prom1:8080"),
+		e2ethanos.DefaultPrometheusImage(),
+	)
+	testutil.Ok(t, err)
+
+	prom2, sidecar2, err := e2ethanos.NewPrometheusWithSidecar(
+		s.SharedDir(),
+		netName,
+		"prom2",
+		defaultPromConfig("ha", 1, "", "", "localhost:9090", "sidecar-prom1:8080"),
```
Comment (prmsrswt) [inline test/e2e/metadata_api_test.go:41]:
```suggestion
		defaultPromConfig("ha", 1, "", "", "localhost:9090", "sidecar-prom2:8080"),
```

**Review by prmsrswt — APPROVED** [review]
LGTM but let's wait for others to review. Awesome work @yeya24!

File: test/e2e/metadata_api_test.go (hunk 5 above)
Comment (yeya24) [reply test/e2e/metadata_api_test.go:41]: Thanks for spotting this!

**Comment by yeya24** [comment]
@prmsrswt Thanks for the detailed review. I have addressed your comments.

File: pkg/metadata/metadatapb/rpc.proto [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+option (gogoproto.goproto_unrecognized_all) = false;
+option (gogoproto.goproto_sizecache_all) = false;
+
+service Metadata {
+  rpc Metadata(MetadataRequest) returns (stream MetadataResponse);
+}
+
+message MetadataRequest {
+  string metric = 1;
+  int32 limit = 2;
```
Comment (bwplotka) [inline pkg/metadata/metadatapb/rpc.proto:29]: What's the limit?

File: pkg/metadata/metadatapb/rpc.proto [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  PartialResponseStrategy partial_response_strategy = 3;
+}
+
+message MetadataResponse {
+  oneof result {
+    /// group for rule groups. It is up to server implementation to decide how many of those to put here within single frame.
```
Comment (bwplotka) [inline pkg/metadata/metadatapb/rpc.proto:35]: Comment should probably start with metadata? something cut off?

**Review by bwplotka — APPROVED** [review]
This is solid work, I think it's good for start TBH 🤗 

I would consider merging this for 0.19. Objections @kakkoyun @squat @prmsrswt ?

File: pkg/metadata/metadatapb/rpc.proto (hunk 6 above)
Comment (yeya24) [reply pkg/metadata/metadatapb/rpc.proto:29]: https://prometheus.io/docs/prometheus/latest/querying/api/#querying-metric-metadata It is used to limit the # of returned metadata.

File: pkg/metadata/metadatapb/rpc.proto (hunk 7 above)
Comment (yeya24) [reply pkg/metadata/metadatapb/rpc.proto:35]: I missed this. Thanks!

**Review by kakkoyun — APPROVED** [review]
🥇

**Comment by bwplotka** [comment]
:scream:

**Pull request merged**